### ISSUE TITLE
[apex] comments on constructor not recognized when the Class has inner class

### ISSUE THREAD

**Issue by MessShawn** [pmd/pmd#1783]
Affects PMD Version: 6.13.0

Rule: [ApexDoc](https://pmd.github.io/pmd-6.14.0/pmd_rules_apex_documentation.html#apexdoc)

Description: comments on constructor not being recognized when the apex class has inner classes

Code Sample demonstrating the issue:
```
/**
 * @description outer class
 */
public class OuterClass {

    private String testStr;

    /**
     * @description constructor
     *
     * @param       testStr     test string
     */
    public OuterClass(String testStr) { 
        this.testStr = testStr;
    }

    /**
     * @description inner class
     */
    public class InnerClass {

    }
}
```

when I run pmd with the rule, it reports that there is Missing ApexDoc Comment on the OuterClass constructor, after I removed the InnerClass definition, it passed the pmd check.

Running PMD through: CLI

Java Version: "1.8.0_25"
OS: Mac Mojave 10.14.4

**Comment by jeffhube** [comment]
@jsotuyod So I think this was working originally, but some changes in 29c5d5b49bee8e25fed0066e37a0bb3bf7f6a4fc may have broke it. If I revert some of the changes like so: https://github.com/pmd/pmd/compare/master...jeffhube:fix-apexdoc , then the new test I added for this case will pass.

It looks like the current approach with the reused list iterator assumes that the nodes will be visited in the same order that they are in the file, which does not seem to be the case. The inner class is visited before the constructor of the outer class, so when we get to the constructor the list iterator has already moved past its doc comment.

The old approach relied on finding the very first token of the class/method/etc declaration (which requires looking for annotations inside `visitEnd`), and then expected the doc comment to be exactly preceding that token, ignoring whitespace. However, I realize that the way I implemented it may not be particularly efficient, so I'll look at improving it and the create a pull request.

Or if anyone has other suggestions, those are welcome too.
